Improve line matching mechanism by introducing line ids

T2L opened the pull request:
This change allows for matching lines that were edited.

T2L committed:
Improve line matching mechanism by introducing line ids

This change allows for matching lines that were edited.

T2L committed:
Address PHPCS errors

T2L committed:
Add test coverage for the new line matching mechanism

T2L commented:
AppVeyor fails due to issues with configuration

@greg-1-anderson, please review

greg-1-anderson commented:
Don't worry about the Appveyor failure; the cause is that they only offer one version of PHP at a time, and they recently switched from 7.3 to 7.4 as their one version. We'd have to add PHP 7.4 support here to bring back Appveyor testing. Probably not hard, but optional.

greg-1-anderson reviewed (changes requested):
Could you better explain how getLineId is working, and how that helps identify and re-assemble the content and comment lines?

T2L commented:
As far as I understand the existing code is not able to match comments with _edited_ lines. Such comments will be lost.

My patch tries to work around this issue by introducing line ids. Let's go with the slightly changed example from tests:

```yaml
# Top comments
top:
  # Top one
  one:
    # Top two
    two: two
# Bottom comments
bottom:
  # Bottom one
  one:
    # Bottom two
    two: 2
```

My code generates ids based on keys (discarding values). Additionally, duplicating keys are taken into account as well. The following ids will be generated:

```
1|top
1|  one
1|    two
1|bottom
2|  one
2|    two
```

This approach (partially) handles this issue:

    > Comments that appear before sections of yaml that are edited may
    > be inadvertently lost. It is recommended to always place comments
    > immediately before identifier lines (i.e. "foo:").

As a result, I was able to preserve all comments in my interactively edited yaml file.

Hope my explanations are clear :)

greg-1-anderson commented:
That explanation is good; could you make it a code comment or part of the documentation, though?

Also, somewhere it would be good to compare/contrast what happens if this remapping is not done. For example, in the existing implementation of this code, I think that maybe the comment "Bottom two" would end up being attached to both "two:" elements.

greg-1-anderson commented:
The README implies that the existing code works in the same way as the new code. Maybe you could move just the test to another PR, so that we can see a failing test?

T2L committed:
Ensure that passed args are valid YAML

T2L committed:
Merge remote-tracking branch 'upstream/master' into patch-1

# Conflicts:
#	tests/TestComments.php

T2L committed:
Add detailed inline comments to the getLineId() method

T2L commented:
hi @greg-1-anderson:
- I've added new PR with - see #4 
- My test YAMLs were not valid (oops). So I've added an extra step to actually validate them
- Added detailed comments to the method

T2L commented:
hi @greg-1-anderson - any updates on this one?

T2L commented:
hi @greg-1-anderson - are you planning to commit this PR?

greg-1-anderson commented on src/Comments.php:
```diff
@@ -17,16 +17,13 @@
- *
- * Comments that appear before sections of yaml that are edited may
- * be inadvertantly lost. It is recommended to always place comments
- * immediately before identifier lines (i.e. "foo:").
```
I don't think that this enhancement entirely obviates the risk described in this comment. I think we should bring this warning back.

T2L replied:
PR has been updated

T2L committed:
Bring back code comments regarding possibility of the data loss

T2L commented:
hi @greg-1-anderson - any chance this would be committed? It's been almost half of a year now :) Just let me know. Don't want to fork but unless this is the only option - I will.

Thanks!

P.S. Hope, you are safe
